### PR TITLE
Remove dotfiles from build files in CMakeLists.txt

### PR DESCRIPTION
**Pull request by MichaelGrupp**
Avoids including unwanted files that could appear in a developer
workspace (e.g. auto-generated files in "dot" folders).

### CODE REVIEW

**Comment by MichaelGrupp** [comment]
Concrete example: I noticed that when configuring the [cquery](https://github.com/cquery-project/cquery) language server for Cartographer. The corresponding vscode extension generates in a `.vscode` folder with header files that mess up the CMake build if the current brute-force glob is used.

Since most IDEs generate some kind of metadata folder in the root dir, this could affect also other developers at some point. Therefore: glob a little bit more specific.

**Comment by MichaelGrupp** [comment]
Any idea why the build only fails on "Jessie"? https://travis-ci.org/googlecartographer/cartographer/builds/340585647

**Comment by gaschler** [comment]
The build fails iff `BUILD_GRPC` is enabled (which is currently tested only on one build server).
`../cartographer_grpc/framework/rpc_handler_interface.h:23:27: fatal error: grpc++/grpc++.h: No such file or directory #include "grpc++/grpc++.h"`

**Comment by MichaelGrupp** [comment]
Ah ok, didn't know it's only on one server. I'll try to reproduce on my machine.

**Comment by MichaelGrupp** [comment]
I just ran this and it worked fine:
```bash
catkin build cartographer --force-cmake -DBUILD_GRPC=true -v --pre-clean > build.txt
grep grpc build.txt > build_grpc_grep.txt
```
The logs show that grpc was built successfully, so at least with this build command it doesn't look like CMake is the issue? You probably use another build config, or there's sth. wrong with the Jessie environment? Hard to say.

[build_grpc_grep.txt](https://github.com/googlecartographer/cartographer/files/1719652/build_grpc_grep.txt)
[build.txt](https://github.com/googlecartographer/cartographer/files/1719653/build.txt)

**Comment by cschuet** [comment]
Drive by comment: I don't think the Jessie build bot has gRPC installed.

**Comment by gaschler** [comment]
I correct myself, our [Jessie build server](https://github.com/googlecartographer/cartographer/blob/master/Dockerfile.jessie) should not compile or require the grpc parts. It builds the default config.

Just an idea, does `cartographer*/*.cc` recurse multiple directories?

**Comment by MichaelGrupp** [comment]
> Just an idea, does cartographer*/*.cc recurse multiple directories?

It has to, otherwise subdirs like `sensor`, `mapping` etc. wouldn't be built. Also my own build would have failed.

**Comment by MichaelGrupp** [comment]
> I correct myself, our Jessie build server should not compile or require the grpc parts. It builds the default config.

Note (to myself): If `-DBUILD_GRPC` is false, the gRPC dirs should be removed from the list by [this logic](https://github.com/magazino/cartographer/blob/3c5358beb03911fa550e97b50fa37c09c80a30ea/CMakeLists.txt#L62). Check.

If **all** build environments don't build gRPC, then this is still strange that it (unintentionally) builds gRPC support **only** on Jessie.

**Comment by MichaelGrupp** [comment]
Just tried to hide my gRPC files in order to reproduce that:
```bash
sudo mv /usr/local/include/grpc /usr/local/include/no-grpc

# this fails (as expected)
catkin build cartographer --force-cmake -DBUILD_GRPC=true -v --pre-clean

# this works (as expected)
catkin build cartographer --force-cmake -DBUILD_GRPC=false -v --pre-clean
```

Another possibility might be that the glob command behaves differently across the CMake versions:
* Jessie: 3.0 (https://packages.debian.org/jessie/cmake) 
* Stretch: 3.7
* Xenial (like my PC): 3.5.1
* Trusty: 2.8.12

Other than that, I have no idea what could be wrong.

**Comment by ojura** [comment]
Some CMake debugging tips:
- you can always do traditional printf-style debugging with message(STATUS ...)
- make a private github fork, go to www.travis-ci.com (not .org!) and start a debug build (this feature is not available for public repos). It provides you with an ssh shell to the debug build VM instance, where you can manually start a travis build and snoop around when it breaks.

**Comment by MichaelGrupp** [comment]
I finally took the time today to continue working on this and came with a better approach (in my opinion). Now, all dotfiles are removed from the glob'd source files in a [separate step](https://github.com/magazino/cartographer/blob/c0329a4093371dd1316948b3395dffe0807bb4e3/CMakeLists.txt#L60) and it's clearer what's going on. More importantly, it builds now on every CI pipeline :tada: 

@gaschler PTAL

File: CMakeLists.txt [hunk 1]
```diff
@@ -56,6 +56,15 @@ install(DIRECTORY cmake DESTINATION share/cartographer/)
 file(GLOB_RECURSE ALL_LIBRARY_HDRS "*.h")
 file(GLOB_RECURSE ALL_LIBRARY_SRCS "*.cc")
 file(GLOB_RECURSE ALL_TESTS "*_test.cc")
+
+# Remove dotfiles/-folders that could potentially pollute the build.
+file(GLOB_RECURSE ALL_DOTFILES ".*/*")
+if (ALL_DOTFILES)
+  list(REMOVE_ITEM ALL_LIBRARY_HDRS ${ALL_DOTFILES})
+  list(REMOVE_ITEM ALL_LIBRARY_SRCS ${ALL_DOTFILES})
+  list(REMOVE_ITEM ALL_TESTS ${ALL_DOTFILES})
```
Comment (gaschler) [inline CMakeLists.txt]: apply remove_item also to ALL_EXECUTABLES?

Comment (MichaelGrupp) [reply CMakeLists.txt]: done

**Comment by gaschler** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user gaschler. Merge queue now has a length of 1.

**Pull request merged**